### PR TITLE
use Factor base class

### PR DESCRIPTION
**Pull request by mark-h-humi**
follow-up to https://github.com/Humi-HR/taxman/pull/75/files#discussion_r1307915904

#75 introduced a base class for tax factors; let's use it whereever we can

image

### CODE REVIEW

File: lib/taxman/taxman2023/t3.rb [hunk 1]
```diff
@@ -16,17 +14,15 @@ class T3
     }.freeze
 
     def initialize(a:, hd:, k2:, tc: nil, k3: 0, tc_offset: 0)
-      @a = a.to_d
-      @hd = hd.to_d
-      @k2 = k2.to_d
-      @k3 = k3.to_d # Other federal non-refundable tax credits
+      super
       @tc = tc&.to_d
       @tc_offset = tc_offset&.to_d
```
Comment (claus-humi) [inline lib/taxman/taxman2023/t3.rb:19]: Should be default 0 if not set.

File: lib/taxman/taxman2023/t4_generic.rb [hunk 2]
```diff
@@ -16,17 +11,14 @@ def initialize(
       tcp: nil,
       tcp_offset: 0
     )
-      @a = a.to_d # Annualized income
-      @hd = hd.to_d
+      super
       @tcp = tcp&.to_d # Personal exemption from provincial TD1, or we use the basic exemption
-      @tcp_offset = tcp_offset.to_d # Offset to personal exemption from provincial TD1, or we use the basic exemption
```
Comment (claus-humi) [inline lib/taxman/taxman2023/t4_generic.rb:22]: Should default to 0

File: lib/taxman/taxman2023/factor.rb [hunk 3]
```diff
@@ -5,7 +5,9 @@ module Taxman2023
   class Factor
     def initialize(**params)
       self.class.params.each do |param|
-        instance_variable_set("@#{param}", params[param].to_d)
+        value = params[param]
+        value = value.to_d unless value.is_a?(String)
```
Comment (claus-humi) [inline lib/taxman/taxman2023/factor.rb:9]: How does `nil.to_d` work?

**Review by claus-humi — CHANGES_REQUESTED** [review]

File: lib/taxman/taxman2023/factor.rb (hunk 3 above)
Comment (mark-h-humi) [reply lib/taxman/taxman2023/factor.rb:9]: it's BigDecimal 0

File: lib/taxman/taxman2023/t4_generic.rb (hunk 2 above)
Comment (mark-h-humi) [reply lib/taxman/taxman2023/t4_generic.rb:22]: yes that will happen in `super`; i did not change the logic

File: lib/taxman/taxman2023/t3.rb (hunk 1 above)
Comment (mark-h-humi) [reply lib/taxman/taxman2023/t3.rb:19]: as this is a refactor i did change any logic

Comment (mark-h-humi) [reply lib/taxman/taxman2023/t3.rb:19]: so basically the subclasses still have initializers only if they do something different that `Factor#initialize`

**Review by claus-humi — APPROVED** [review]
Looks more consistent now.

It's a weird part of the codebase and I hope that we starting 2024 won't have to deal with all these `nil` values as there should be actual inputs into these fields.

**Pull request merged**